Add stocktaking button to my_equipment view

Dragerus opened the pull request:
New button in 'my equipment' view (at the end of each table row). On click adds stocktaking tag to asset. Can be disabled in settings/via environmental variable

Dragerus committed:
Add stocktaking button to my_equipment view

New button in 'my equipment' view (at the end of each table row). On click adds stocktaking tag to asset. Can be disabled in settings/via environmental variable

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -28,6 +31,51 @@ def user_licence(self, item):
+        if not 'asset_id' in request.POST:
+            return HttpResponseBadRequest()
+
+        asset = BackOfficeAsset.objects.get(pk=request.POST['asset_id'])
```
use [`get_object_or_404`](https://docs.djangoproject.com/en/1.9/topics/http/shortcuts/#get-object-or-404) from `django.shortcuts`

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -28,6 +31,51 @@ def user_licence(self, item):
+    
+    def post(self, request, *args, **kwargs):
+        if not 'asset_id' in request.POST:
+            return HttpResponseBadRequest()
```
IMO this block is redundant - because [this line](https://github.com/allegro/ralph/pull/2602/files#diff-8684faeb9511ab3569426359d236c9f7R18) care of this by raising `Http404`

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -28,6 +31,51 @@ def user_licence(self, item):
+        asset.tags.add(settings.INVENTORY_TAG)
+        asset.save()
+
+        return render(
```
why not `return super().post(request, *args, **kwargs)`?

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -28,6 +31,51 @@ def user_licence(self, item):
+            context=self.get_context_data()
+        )
+
+    def get_context_data(self, **kwargs):
```
Are you missing add some extra arguments to context? Or this block is not necessary?

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -28,6 +31,51 @@ def user_licence(self, item):
+    template_name = 'ralphuser/inventory_tag_confirmation.html'
+
+    def get_context_data(self, **kwargs):
+        print(self.request.method)
```
rm / dd

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -28,6 +31,51 @@ def user_licence(self, item):
+class InventoryTagView(RalphBaseTemplateView):
+    template_name = 'ralphuser/inventory_tag.html'
+    
+    def post(self, request, *args, **kwargs):
```
change it to `def post(self, request, asset_id, *args, **kwargs)` then you can access directly instead of `request.POST['asset_id']`

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -28,6 +31,51 @@ def user_licence(self, item):
+
+        asset = BackOfficeAsset.objects.get(pk=request.POST['asset_id'])
+        if not asset.user.username == request.user.username:
+            return HttpResponseForbidden()
```
IMO `dispatch` method is better to do this

ar4s commented on src/ralph/accounts/urls.py:
```diff
@@ -1,7 +1,7 @@
 from django.contrib.auth.decorators import login_required
 
-from ralph.accounts.views import CurrentUserInfoView, UserProfileView
+from ralph.accounts.views import CurrentUserInfoView, UserProfileView, InventoryTagView, InventoryTagConfirmationView
```
pep8

Dragerus committed:
WIP append date to inventory tag

Dragerus committed:
WIP display message for tagged assets

Issue RAS-771:
User didn't like fact there was just empty field in table.

Dragerus committed:
WIP add "no" option in table

User has now option to deny that he has assigned asset. They will
receive a message with further instructions.

Dragerus committed:
WIP add functional tests

Dragerus committed:
WIP beautify confirmation form. Add cancel button

Issue: RAS-772
Form didn't have cancel button and was considered ugly.

Dragerus committed:
WIP move success information to my_equipment view

Issue: RAS-770

xliiv commented on src/ralph/accounts/views.py:
```diff
@@ -28,9 +38,70 @@ def user_licence(self, item):
+        if not asset.user_id == request.user.id:
+            return HttpResponseForbidden()
+
+        date_tag = None
```
this is useless?

Dragerus replied:
actually I needed to define this variable for later use

xliiv replied:
I meant that this is enough
`date_tag = settings.INVENTORY_TAG + '_' + date.today().isoformat()`
AFAIK

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -81,12 +85,23 @@ def post(self, request, *args, **kwargs):
+    def dispatch(self, request, *args, **kwargs):
+        try:
+            if request.COOKIES.get('tag_success'):
+                messages.success(request, _('Asset successfully tagged'))
```
move this to `InventoryTagView.post` after `asset.save` -- django's message framework handle of all stuff to display message to user in next request

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -1,10 +1,15 @@
 # -*- coding: utf-8 -*-
+from _ast import arg
```
dd

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -1,10 +1,15 @@
 from django.shortcuts import get_object_or_404
 from django.utils.translation import ugettext_lazy as _
+from django.views.generic import View
+from pip import req
```
dd

Dragerus committed:
WIP improve asset missing handling

Missing assets now have their own tag

Dragerus committed:
WIP add stocktaking switch for warehouses and regions

Issue: RAS-769
Privileged user can now start stocktaking for
selected warehouse or region. End-user will have
option of self-stocktaking if they are assigned
to specific region or they have assigned assets
associated with specified warehouse.

ar4s commented on src/ralph/accounts/templates/ralphuser/inventory_tag_confirmation.html:
```diff
@@ -7,17 +7,23 @@
-          class="button small success left" autocomplete="off" />
-      <a href="javascript:history.back()" class="button small cancel left">
+      <input type="hidden" name="answer" value={{ answer }} />
+      <a href="javascript:history.back()" class= "button small alert left">
```
IMO `reverse('current_user_info')` is better choice -- i don't trust JS

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -55,19 +55,17 @@ def get_context_data(self, **kwargs):
-    http_method_names = ['post']
-
     def post(self, request, *args, **kwargs):
+        response = HttpResponseRedirect(reverse('current_user_info'))
```
where you use response variable (excluding return statement)?

Dragerus replied:
nowhere, remained after playing with cookies

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -76,30 +74,33 @@ def post(self, request, *args, **kwargs):
-            date_tag
-        )
-        asset.save()
+        if request.POST['answer'] == 'yes':
```
IMO you should split this two block of code (inside `if`s) to two separeted methods like `_post_yes(self, request, asset)` and `_post_no(self, request, asset)`. Benefits? Easliy to maintence, testing, debugging and read.

ar4s commented on src/ralph/accounts/views.py:
```diff
@@ -39,17 +117,28 @@ def get_context_data(self, **kwargs):
         if settings.MY_EQUIPMENT_REPORT_FAILURE_URL:
             asset_fields += ['report_failure']
+
+        if BackOfficeAsset.objects.filter(
```
1. use `.exists()` instead of `.count() > 0`
2. move conditions to variables ex. `warehouse_stocktaking_enabled`, `region_stocktaking_enabled`

xliiv commented on src/ralph/settings/base.py:
```diff
@@ -317,6 +317,11 @@ def os_env_true(var, default=''):
+INVENTORY_TAG = os.environ.get('INVENTORY_TAG', 'INV')
+INVENTORY_TAG_USER = os.environ.get('INVENTORY_TAG_USER', 'INV_CONF')
+INVENTORY_TAG_APPEND_DATE = os_env_true('INVENTORY_TAG_APPEND_DATE', '1')
+INVENTORY_TAG_MISSING = os.environ.get('INVENTORY_TAG_MISSING', 'INV_MISSING')
```
Could you add comments describing these vars?

Dragerus replied:
Each one of them?

Dragerus committed:
WIP beautify my_equipment view. Fix stocktaking logic

Made buttons in my_equipment view less ugly and fixed minor
logic bug.

Dragerus committed:
WIP apply recommended changes

Dragerus committed:
WIP update tests

ghost commented:
:ok:

ar4s commented:
:ok:

vi4m commented on src/ralph/accounts/admin.py:
```diff
@@ -136,6 +136,38 @@ def escape_param(p):
+                '<div class="small-12 columns label alert">missing</div>'
+            )
+        else:
+            return '<a class="small-6 columns label success" href="{}">{}</a>' \
```
named params would be better: `'<a class="small-6 columns label success" href="{confirm_url}">{confirm_caption}</a>'.format(confirm_url=..., confirm_caption=...)`

vi4m commented on src/ralph/accounts/models.py:
```diff
@@ -32,6 +32,7 @@ class Region(PermissionsForObjectMixin, NamedMixin):
         choices=Country(),
         default=Country.pl.id,
     )
+    stocktaking_enabled = models.BooleanField(default=False)
```
help text missing, describe what that means
